crypto/x509/x_attrib.c

```diff
@@ -56,3 +56,221 @@ X509_ATTRIBUTE *X509_ATTRIBUTE_create(int nid, int atrtype, void *value)
     return NULL;
 }
+
+static int print_hex(BIO *out, unsigned char *buf, int len)
+{
+    int i;
+    for (i = 0; i < len; i++) {
+        if (BIO_printf(out, "%02X ", buf[i]) <= 0) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static int ASN1_INTEGER_print_bio(BIO *bio, const ASN1_INTEGER *num)
+{
+    BIGNUM *num_bn;
+    int result = 0;
+    char *hex;
+
+    num_bn = ASN1_INTEGER_to_BN(num, NULL);
+    if (num_bn == NULL)
+        return -1;
+    if ((hex = BN_bn2hex(num_bn))) {
+        result = BIO_write(bio, "0x", 2) > 0;
+        result = result && BIO_write(bio, hex, strlen(hex)) > 0;
+        OPENSSL_free(hex);
+    }
+    BN_free(num_bn);
+
+    return result;
+}
+
+static int print_oid (BIO *out, ASN1_OBJECT *oid) {
+    const char *ln;
+    char objbuf[80];
+
+    if (OBJ_obj2txt(objbuf, sizeof(objbuf), oid, 1) <= 0) {
+        return 0;
+    }
+    ln = OBJ_nid2ln(OBJ_obj2nid(oid));
+    return (ln != NULL)
+        ? BIO_printf(out, "%s (%s)", objbuf, ln)
+        : BIO_printf(out, "%s", objbuf);
+}
+
+int print_attribute_value(BIO *out, int obj_nid, const ASN1_TYPE *av, int indent)
+{
+    const char *ln;
+    char objbuf[80];
+    ASN1_STRING *str;
+    unsigned char *value;
+    X509_NAME *xn = NULL;
+    int64_t int_val;
+
+    // This switch-case is only for syntaxes that are not encoded as a single
+    // primitively-constructed value universal ASN.1 type.
+    switch (obj_nid) {
+    case NID_undef: break; // Unrecognized OID.
+    // Attribute types with DN syntax.
+    case NID_member:
+    case NID_roleOccupant:
+    case NID_seeAlso:
+    case NID_manager:
+    case NID_documentAuthor:
+    case NID_secretary:
+    case NID_associatedName:
+    case NID_dITRedirect:
+    case NID_owner:
+        value = av->value.sequence->data;
+        if ((xn = d2i_X509_NAME(NULL, (const unsigned char**)&(av->value.sequence->data), av->value.sequence->length)) == NULL) {
+            BIO_puts(out, "(COULD NOT DECODE DISTINGUISHED NAME)\n");
+            return 0;
+        }
+        // d2i_ functions increment the ppin pointer. See doc/man3/d2i_X509.pod.
+        // This resets the pointer. We don't want to corrupt this value.
+        av->value.sequence->data = value;
+        if (X509_NAME_print_ex(out, xn, indent, 0) <= 0) {
+            return 0;
+        }
+        X509_NAME_free(xn);
+        return 1;
+
+    default: break;
+    }
+
+    switch (av->type) {
+    case V_ASN1_BOOLEAN:
+        if (av->value.boolean) {
+            return BIO_printf(out, "%*sTRUE", indent, "");
+        } else {
+            return BIO_printf(out, "%*sFALSE", indent, "");
+        }
+
+    case V_ASN1_INTEGER:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        if (ASN1_INTEGER_get_int64(&int_val, av->value.integer) > 0) {
+            return BIO_printf(out, "%ld", int_val);
+        } else {
+            return ASN1_INTEGER_print_bio(out, str);
+        }
```

Comment:
This series of cases is a bit confusing to me.  You have a case statement here that identifies the type of value that a given ASN1 string is, but in the event it can't be decoded, we just print the string.  Shouldn't that be an error condition?  Or are there cases in which this is normal?

Reply:
I feel like you are looking at an old version of the file here. I had a mistake in here earlier, which I fixed. (`str` was never set.) What my current code does (and what my original intent was) is print the `INTEGER` as a decimal if it can fit in `int64_t`, but in hexadecimal otherwise. I will mark this as resolved unless you look at my newest code and believe it to still be a problem.